File: structures/lists/linked_list.js
class Node {
  constructor(data = null) {
    this.data = data;
    this.next = null;
  }
  toString() {
    return `${this.data} -> ${this.next}`
  }
}

class LinkedList {
  constructor(head = null) {
    this.head = head;
  }
  is_empty() {
    return this.head === null;
  }
  size() {
    let current = this.head;
    let length = 0
    while (current) {
      length += 1;
      current = current.next;
    }
    return length
  }
  add(data) {
    // Prepending
    let node = new Node(data);
    node.next = this.head;
    this.head = node;
  }
  search(key) {
    let current = this.head;
    while (current) {
      if (current.data === key) {
        return current;
      } else {
        current = current.next
      }
    }
    return null
  }
  insert(data, index) {
    if (index === 0) {
      this.add(data);
      return true;
    } else if (index < 0) {
      return false;
    }
    let current = this.head;
    let new_node = new Node(data);
    while (index > 0) {
      if (!current) return false;
      current = current.next;
      index -= 1;
    }
    let prev = current;
    let next = current.next;
    prev.next = new_node;
    new_node.next = next;
  }
  remove(key) {
    current = this.head;
    previous = null;
    found = false;
    while (current && !found) {
      if (current.data === key && current === this.head) {
        this.head = current.next;
        found = true;
      } else if (current.data === key) {
        previous.next = current.next;
        found = true;
      } else {
        previous = current;
        current = current.next;
      }
    }
    return current.data
  }
  toString() {
    return `${this.head}`
  }
}


const n1 = new Node(0);
console.log(`N1: ${n1}`);

const l1 = new LinkedList();
console.log(`LinkedList: ${l1}`);

console.log(`Is Empty: ${l1.is_empty()}`);
l1.head = n1;
console.log(`Is Empty: ${l1.is_empty()}`);

l1.add(10);
l1.add(20);
l1.add(30);
l1.add(40);
l1.add(60);
l1.add(70);
l1.add(80);
l1.add(90);
l1.add(100);

console.log(`LinkedList: ${l1}`);
console.log(`Size: ${l1.size()}`);

l1.insert(50, 5);
console.log(`LinkedList: ${l1}`);
console.log(`Size: ${l1.size()}`);

console.log(l1.size());

let n = l1.search(20);
console.log(n);

File: structures/lists/linked_list.py

class Node:
    """
    An object for storing a single node of a Linked List
    Models two attributes - data and the link to the next node
    """
    data = None
    next = None
    def __init__(self, data):
        self.data = data
    def __repr__(self):
        return f"{self.data}  ->  {self.next}"

class LinkedList:
    """
    Singly linked list
    """
    def __init__(self):
        self.head = None

    def is_empty(self):
        return self.head is None

    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.next
        return count

    def add(self, data):
        """
        Adds a new node to the Linked List, happens in constant time. (Pre-pend)
        """
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
    
    def search(self, key):
        """
        Finds the node with the data that matches the key
        Returns the node or None if not found
        Linear Time O(n)
        """
        current = self.head
        while current:
            if current.data == key:
                return current
            else:
                current = current.next
        return current

    def insert(self, data, index):
        """
        Inserts a new node containing data at position
        Insertion takes constant O(1) but finding position takes O(n)
        """
        if index == 0:
            self.add(data)
        if index > 0:
            new = Node(data)
            position = index
            current = self.head
            while position > 1:
                current = current.next
                position -= 1
            prev = current
            next = current.next
            prev.next = new
            new.next = next

    def remove(self, key):
        """
        Removes a key from the list
        Takes Linear time
        """
        current = self.head
        previous = None
        found = False
        while current and not found:
            if current.data == key and current is self.head:
                found = True
                self.head = current.next
            elif current.data == key:
                found = True
                previous.next = current.next
            else:
                previous = current
                current = current.next
        return current.data


# Add print statement that shows next and previous pointers
# def remove at index
# def read at index

l1 = LinkedList()
n1 = Node(0)

l1.head = n1
l1.add(10)
l1.add(20)
l1.add(30)
l1.add(40)
l1.add(50)
l1.add(60)
l1.add(70)
l1.add(80)
l1.add(90)
l1.add(100)

n = l1.search(20)
print(n)